### PR TITLE
ddl: support alter index for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

Problem Summary:

This PR supports alter multiple indexes in one DDL statement.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -352,8 +357,13 @@ func onAlterIndexVisibility(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64,
 	if err != nil || tblInfo == nil {
 		return ver, errors.Trace(err)
 	}
-	idx := tblInfo.FindIndexByName(from.L)
-	idx.Invisible = invisible
+
+	if job.MultiSchemaInfo != nil && job.MultiSchemaInfo.Revertible {
+		job.MarkNonRevertible()
+		return updateVersionAndTableInfo(d, t, job, tblInfo, false)
```
Comment (zimulala) [inline ddl/index.go:363]: Do we need use `updateVersionAndTableInfoWithCheck` instead of `updateVersionAndTableInfo`?

Comment (tangenta) [reply ddl/index.go:363]: The `TableInfo` is not changed. Is it really necessary?

Comment (zimulala) [reply ddl/index.go:363]: It is not necessary

File: ddl/multi_schema_change_test.go [hunk 2]
```diff
@@ -926,6 +1031,17 @@ func TestMultiSchemaChangeAdminShowDDLJobs(t *testing.T) {
...
 
+func TestMultiSchemaChangeAlterIndexVisibility(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test;")
+	tk.MustExec("create table t (a int, b int, index idx(b));")
+	tk.MustExec("alter table t add index idx2(a), alter index idx visible;")
+	tk.MustQuery("select * from t use index (idx, idx2);").Check(testkit.Rows( /* no rows */ ))
+	tk.MustGetErrCode("alter table t drop column b, alter index idx invisible;", errno.ErrKeyDoesNotExist)
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/multi_schema_change_test.go:1042]: OK.

**Review by zimulala — APPROVED** [review]
LGTM

File: ddl/index.go [hunk 3]
```diff
@@ -338,11 +338,16 @@ func onRenameIndex(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
...
-func validateAlterIndexVisibility(indexName model.CIStr, invisible bool, tbl *model.TableInfo) (bool, error) {
-	if idx := tbl.FindIndexByName(indexName.L); idx == nil {
+func validateAlterIndexVisibility(ctx sessionctx.Context, indexName model.CIStr, invisible bool, tbl *model.TableInfo) (bool, error) {
+	var idx *model.IndexInfo
+	if idx = tbl.FindIndexByName(indexName.L); idx == nil || idx.State != model.StatePublic {
 		return false, errors.Trace(infoschema.ErrKeyNotExists.GenWithStackByArgs(indexName.O, tbl.Name))
-	} else if idx.Invisible == invisible {
-		return true, nil
+	}
+	if ctx == nil || ctx.GetSessionVars() == nil || ctx.GetSessionVars().StmtCtx.MultiSchemaInfo == nil {
```
Comment (wjhuang2016) [inline ddl/index.go:346]: in what situation `ctx == nil || ctx.GetSessionVars() == nil`?

**Review by wjhuang2016 — APPROVED** [review]

**Comment by hawkingrei** [comment]
@tangenta 
```
    tiflash_replica_test.go:189: 
        	Error Trace:	tiflash_replica_test.go:189
        	Error:      	Not equal: 
        	            	expected: "[ddl:8200]ALTER table replica for tables in system database is currently unsupported"
        	            	actual  : "[planner:1142]ALTER command denied to user 'root'@'%' for table 'advisory_locks'"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-[ddl:8200]ALTER table replica for tables in system database is currently unsupported
        	            	+[planner:1142]ALTER command denied to user 'root'@'%' for table 'advisory_locks'
        	Test:       	TestSetTableFlashReplicaForSystemTable
```

**Comment by tangenta** [comment]
```
-[ddl:8200]ALTER table replica for tables in system database is currently unsupported
+[planner:1142]ALTER command denied to user 'root'@'%' for table 'advisory_locks'
        	Test:       	TestSetTableFlashReplicaForSystemTable
        	Messages:   	sem enabled: true
```

I don't understand why the SEM is enabled after this PR.

**Pull request merged**